### PR TITLE
Adding template for block test.

### PR DESCRIPTION
**Pull request by moniika**
#### Refactors tests under block-plus-minus to fit template structure:
-  [x] if.mocha.js
-  [x] list_create.mocha.js
-  [x] text_join.mocha.js

### Additional Information
Refactor of `block-plus-minus/procedures.mocha.js` to follow in separate PR

### CODE REVIEW

File: plugins/dev-create/templates/block/template/test/block_test.mocha.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {assert} from 'chai';
+import {sinon} from 'sinon';
```
Comment (samelhusseini) [inline plugins/dev-create/templates/block/template/test/block_test.mocha.js:8]: Does that mean sinon needs to be added to the block template's package.json?

Comment (samelhusseini) [reply plugins/dev-create/templates/block/template/test/block_test.mocha.js:8]: I guess same for chai.

Comment (samelhusseini) [inline plugins/dev-create/templates/block/template/test/block_test.mocha.js]: Does this mean sinon and chai need to be added to the block template's package.json?

File: plugins/dev-create/templates/block/template/test/block_test.mocha.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    test('xmlToBlock', function() {
+      const block = Blockly.Xml.domToBlock(Blockly.Xml.textToDom(
+          '<block type="block_template"/>'
+      ), this.workspace);
+      // TODO
+    });
+
+    // TODO see block_test
```
Comment (samelhusseini) [inline plugins/dev-create/templates/block/template/test/block_test.mocha.js]: What is this referring to?

File: plugins/dev-create/templates/block/template/test/block_test.mocha.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      const expectedXml = ''; // TODO
+      const xml = Blockly.Xml.domToText(Blockly.Xml.blockToDom(this.block));
+      assert.equal(xml, expectedXml);
+    });
+
+    test('xmlToBlock', function() {
+      const block = Blockly.Xml.domToBlock(Blockly.Xml.textToDom(
+          '<block type="block_template"/>'
+      ), this.workspace);
+      // TODO
```
Comment (samelhusseini) [inline plugins/dev-create/templates/block/template/test/block_test.mocha.js]: More explicit TODO, what should someone creating a new block do?

File: plugins/dev-create/templates/block/template/test/block_test.mocha.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    // assert.exists(this.block.inputList, 'Has inputList');
+    // assert.equal(this.block.inputList.length, 3);
+    // assert.equal(this.block.getInput('IF').connection.check_.length, 1);
+    // assert.equal(this.block.getInput('IF').connection.check_[0], 'Boolean');
+    // assert.exists(this.block.onchangeWrapper_, 'Has onchange handler');
+  });
+
+  suite('blockToCode', function() {
+    // TODO comment out supported generators
+    const supportedGenerators = [
```
Comment (samelhusseini) [inline plugins/dev-create/templates/block/template/test/block_test.mocha.js]: I like this.

**Review by samelhusseini — COMMENTED** [review]
This looks good, but do we want to test this out in the blocks_plus_minus package first before moving it into the template?

File: plugins/dev-create/templates/block/template/test/block_test.mocha.js (hunk 2 above)
Comment (moniika) [reply plugins/dev-create/templates/block/template/test/block_test.mocha.js]: This is more a note for myself block_test in core. I plan on removing it.

File: plugins/block-plus-minus/test/if.mocha.js [hunk 5]
```diff
@@ -27,134 +37,201 @@ suite('if & ifelse', () => {
...
+
+      test('Trivial', function() {
+        const xml =
+            Blockly.Xml.domToPrettyText(Blockly.Xml.blockToDom(this.block));
+        assertXmlMatch(xml, 'controls_if');
+      });
+
+      test('No else', function() {
+        this.block.plus();
+        this.block.plus();
```
Comment (moniika) [inline plugins/block-plus-minus/test/if.mocha.js]: I'm not sure if this is better or if I should instead do round trip with `domToBlock(textToDom(` like in 'With else' test

File: plugins/dev-tools/src/common_test_helpers.mocha.js [hunk 6]
```diff
@@ -6,26 +6,76 @@
...
     testCall = (testCase.only ? test.only : testCall);
-    testCall(testCase.title, testFn(testCase));
+    testCall(testCase.title, createTestCallback(testCase));
+  });
+}
+
+/**
+ * Runs provided test suite.
+ * @param {Array<TestSuite<T>>} testSuites The test suites to run.
+ * @param {function(TestSuite<T>):function(TestCase<T>):Function
```
Comment (moniika) [inline plugins/dev-tools/src/common_test_helpers.mocha.js]: should be instead `{function(TestSuite<T>):function(T):Function}`

Comment (samelhusseini) [reply plugins/dev-tools/src/common_test_helpers.mocha.js]: Yeah I would add parentheses around that second function for clarity.

**Comment by samelhusseini** [comment]
https://github.com/google/blockly-samples/pull/279

**Comment by moniika** [comment]
> #279

Will merge into this PR after that goes in.

**Comment by samelhusseini** [comment]
PR is in if you wanted to merge and try again.

File: plugins/eslint-config/index.js [hunk 7]
```diff
@@ -82,6 +82,12 @@ module.exports = {
   overrides: [
     {
       files: ['**/*.mocha.js'],
+      settings: {
+        jsdoc: {
+          mode: 'closure',
```
Comment (moniika) [inline plugins/eslint-config/index.js:87]: I'm not sure whether this causes any unwanted side effects. 
It is needed for @template tag support

Comment (moniika) [reply plugins/eslint-config/index.js:87]: I could also set this config for all the files. This change here only changes the mode for mocha files.

File: plugins/dev-tools/src/block_test_helpers.mocha.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+export function CodeGenerationTestCase() {}
+CodeGenerationTestCase.prototype = new TestCase();
+/**
+ * @type {string} The expected code.
+ */
+CodeGenerationTestCase.prototype.expectedCode = '';
+/**
+ * @type {number|undefined} The expected inner order.
+ */
+CodeGenerationTestCase.prototype.expectedInnerOrder = undefined;
```
Comment (samelhusseini) [inline plugins/dev-tools/src/block_test_helpers.mocha.js:33]: No harm in setting this to `undefined`, but not sure if it's necessary?

**Review by samelhusseini — APPROVED** [review]

**Pull request merged**